Support numbers in GelfMessage.AdditonalFields

Andreas-Hjortland opened the pull request:
This fixes #17 If an input object is numeric we will treat is as a number in json as well

I also created pull request #19 which takes the content of this and extends the supported `ValueTuple` types in `BeginScope` with all the numeric types of C#

Commit:
Support numbers in GelfMessage.AdditonalFields

Andreas-Hjortland commented:
I am not sure why the build failed. It works fine on my machine running `docker-compose -f docker-compose.ci.build.yml -f docker-compose.yml up --abort-on-container-exit`

![image](https://user-images.githubusercontent.com/2162904/41125323-6b929898-6aa4-11e8-9166-47f650d75220.png)

mattwcole commented:
Nice change, much appreciated! I'll have to have a look at Travis and hopefully get it merged and released this weekend. The whole point of building in Docker in the first place was to make sure the local build is identical to the one on the server. Perhaps we have a flaky test.

mattwcole commented:
I have been able to reproduce the issue locally both within Visual Studio and from the command line using compose. The graylog logs show:

`graylog_1        | 2018-06-09 12:59:19,001 ERROR: org.graylog2.indexer.messages.Messages - Failed to index [1] messages. Please check the index error log in your web interface for the reason. Error: One or more of the items in the Bulk request failed, check BulkResult.getItems() for more information.`

Navigating to http://localhost:9000/system/indices/failures shows the detail:

`{"type":"mapper_parsing_exception","reason":"failed to parse [baz]","caused_by":{"type":"number_format_exception","reason":"For input string: \"baz\""}}`

I'm not sure why the build passed locally for you. Maybe try re-creating the containers with by adding the `--force-recreate` flag to the compose command?

mattwcole commented:
I had a little look into why it might be failing, and I think it is because we use the `baz` field as a string in other tests. From [the docs](http://docs.graylog.org/en/2.4/pages/indexer_failures.html#mapperparsingexception):

_This can be triggered by for example sending GELF messages with different field types or extractors trying to write strings without converting them to numeric values first. The recommended solution is to actively decide on field types. If you sent in a field like http_response_code with a numeric value then you should never change that type in the future._

Also apologies for the merge conflicts, we have just added support for GELF over HTTP.

mattwcole commented on src/Gelf.Extensions.Logging/UdpGelfClient.cs:
```diff
@@ -46,13 +46,36 @@ await _udpClient.SendAsync(messageChunk, messageChunk.Length, _options.Host, _op
+                else
+                {
+                    messageJson[$"_{field.Key}"] = field.Value?.ToString();
+                }
```
I don't think we need `IsNumeric()` here. `JToken` will handle string/number formatting for us. We just need a null check.

```csharp
foreach (var field in message.AdditionalFields.Where(f => f.Value != null))
{
    messageJson[$"_{field.Key}"] = JToken.FromObject(field.Value);
}
```

Andreas-Hjortland replied:
I am aware, but I was afraid that would break for non string values like booleans which should be stringified and not serialized as json

mattwcole replied:
Great point, good job you are implementing this and not me!

Andreas-Hjortland commented:
Thanks... It might be that I didn't catch that since I ran the test in isolation to improve performance does the tests run in a randomized order in docker? If so that might explain why it even worked in docker for me

Andreas-Hjortland commented:
I will review and try to fix the issues this afternoon

mattwcole commented:
I don't think the tests run in a randomised order. When you ran the suite from the command line however, the containers will have kept their state from when you ran the single test, meaning the index will have been set up with `baz` as a number field. When the tests fail, it looks like `baz` is initially set up as a text field. That could cause different behaviour. Recreating your containers and running your new test last should reproduce the problem. That's my theory anyway.

Andreas-Hjortland commented:
That makes sense. I will fix it today

Andreas-Hjortland committed:
Merge branch 'dev' into 17_numeric_additional_fields

Andreas-Hjortland committed:
Fix issue with failing tests

The tests failed because the name 'baz' was already used as a string
value which was indexed by graylog

Andreas-Hjortland committed:
Move number serialization to GelfMessageExtensions

Andreas-Hjortland commented:
I managed to make it work now. I changed the numeric key from "baz" to "quux" as that was used for a numeric value.

Do you want me to squash the changes to a single commit?

mattwcole commented:
Brilliant thanks, don't worry about squashing the commits, I'll do that when I merge.
